cpp/src/arrow/util/logging.h

```diff
@@ -202,48 +169,81 @@ class ARROW_EXPORT ArrowLog : public ArrowLogBase {
  private:
   ARROW_DISALLOW_COPY_AND_ASSIGN(ArrowLog);
 
+  std::ostream* Stream() override;
+
   // Hide the implementation of log provider by void *.
   // Otherwise, lib user may define the same macro to use the correct header file.
   void* logging_provider_;
   /// True if log messages should be logged and false if they should be ignored.
   bool is_enabled_;
 
   static ArrowLogLevel severity_threshold_;
-
- protected:
-  std::ostream& Stream() override;
 };
 
-// This class make ARROW_CHECK compilation pass to change the << operator to void.
-// This class is copied from glog.
-class ARROW_EXPORT Voidify {
- public:
-  Voidify() {}
-  // This has to be an operator with a precedence lower than << but
-  // higher than ?:
-  void operator&(ArrowLogBase&) {}
-};
+struct ARROW_EXPORT InterceptComparison {
+  template <typename NotBinaryOrNotPrintable>
+  static ArrowLogBase&& PrintOperands(const NotBinaryOrNotPrintable&,
+                                      ArrowLogBase&& log) {
+    return std::move(log);
+  }
 
-namespace detail {
+  template <typename Lhs, typename Rhs>
+  struct BoundLhsRhs {
+    explicit constexpr operator bool() const { return false; }
+    const Lhs& lhs;
+    const Rhs& rhs;
+  };
+
+  template <typename Lhs, typename Rhs>
+  static auto PrintOperands(const BoundLhsRhs<Lhs, Rhs>& bound, ArrowLogBase&& log)
+      -> decltype(std::move(log) << bound.lhs << bound.rhs) {
+    return std::move(log) << "\n  left:  " << bound.lhs << "\n  right: " << bound.rhs
+                          << "\n";
+  }
 
-/// @brief A helper for the nil log sink.
-///
-/// Using this helper is analogous to sending log messages to /dev/null:
-/// nothing gets logged.
-class NullLog {
- public:
-  /// The no-op output operator.
-  ///
-  /// @param [in] t
-  ///   The object to send into the nil sink.
-  /// @return Reference to the updated object.
-  template <class T>
-  NullLog& operator<<(const T& t) {
-    return *this;
+  template <typename Lhs>
+  struct BoundLhs {
+    template <typename Rhs>
+    BoundLhsRhs<Lhs, Rhs> operator==(const Rhs& rhs) && {
+      return {lhs, rhs};
+    }
+
+    template <typename Rhs>
+    BoundLhsRhs<Lhs, Rhs> operator!=(const Rhs& rhs) && {
+      return {lhs, rhs};
+    }
+
+    template <typename Rhs>
+    BoundLhsRhs<Lhs, Rhs> operator>(const Rhs& rhs) && {
+      return {lhs, rhs};
+    }
+
+    template <typename Rhs>
+    BoundLhsRhs<Lhs, Rhs> operator>=(const Rhs& rhs) && {
+      return {lhs, rhs};
+    }
+
+    template <typename Rhs>
+    BoundLhsRhs<Lhs, Rhs> operator<(const Rhs& rhs) && {
+      return {lhs, rhs};
+    }
+
+    template <typename Rhs>
+    BoundLhsRhs<Lhs, Rhs> operator<=(const Rhs& rhs) && {
+      return {lhs, rhs};
+    }
```

Comment:
Why all these operators? They don't look like they perform the advertised operations.

Reply:
The assert condition is tested [without decoration](https://github.com/apache/arrow/pull/8735/files#diff-e4af9b24769d2da2516cca1deb194f67f3fdbf692e6486c0e7149279af01cfb6R83). This avoids some compiler warnings and suspicion that destructuring might slow down the fast path (the check passes).

Destructuring operates on another copy of the condition expression for printing purposes only.

It relies on the slightly higher precedence of `<=` compared to the other comparison operators. After macro expansion, the [intercepted comparison](https://github.com/apache/arrow/pull/8735/files#diff-e4af9b24769d2da2516cca1deb194f67f3fdbf692e6486c0e7149279af01cfb6R85) is as follows:
```
InterceptComparison() <= x > y
((InterceptComparison() <= x) > y)  // parentheses for clarity
```
The inner parenthesis captures `x`, then the outer parenthesis captures `y`

Reply:
Same question then :-)

Reply:
This is the approach used by the [Catch2](https://github.com/catchorg/Catch2) test framework's assertion macro. I'm not aware of another way to destructure a comparison's operands

Reply:
Ok, but destructuring in the first place shouldn't be required. I guess you want `DCHECK(a == b)` to function exactly like `DCHECK_EQ(a, b)`, but that doesn't sound required to me. 

Otherwise, please at least explain all this in comments, because it's extremely cryptic at first sight.